scala/src/main/org/apache/spark/streaming/api/kafka/ParallelKafkaDataFetcherDStream.scala

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) Microsoft. All rights reserved.
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+package org.apache.spark.streaming.kafka
+
+import org.apache.spark.storage.StorageLevel
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+import scala.reflect.ClassTag
+import scala.collection.JavaConversions._
+
+import java.util.UUID
+import java.util.concurrent._
+
+import kafka.common.TopicAndPartition
+import kafka.message.MessageAndMetadata
+import kafka.serializer.Decoder
+
+import org.apache.spark.util.ThreadUtils
+import org.apache.spark.rdd.RDD
+import org.apache.spark.streaming.api.csharp.{RddPreComputeRecord, RddPreComputeCache, CSharpDStream}
+import org.apache.spark.streaming.{StreamingContext, Time}
+import org.apache.spark.streaming.dstream._
+import org.apache.spark.streaming.scheduler.StreamInputInfo
+
+/**
+ * input stream that directly pulls messages from a Kafka Broker and specific offset.
+ * It will use separate threads to pull message from Kafka and call the readFunc to process the data,
+ * the result is put in a queue for further consume.
+ *
+ * For an easy interface to Kafka-managed offsets,
+ *  see {@link org.apache.spark.streaming.kafka.KafkaCluster}
+ * @param kafkaParams Kafka <a href="http://kafka.apache.org/documentation.html#configuration">
+ * configuration parameters</a>.
+ *   Requires "metadata.broker.list" or "bootstrap.servers" to be set with Kafka broker(s),
+ *   NOT zookeeper servers, specified in host1:port1,host2:port2 form.
+ * @param fromOffsets per-topic/partition Kafka offsets defining the (inclusive)
+ *  starting point of the stream
+ * @param messageHandler function for translating each message into the desired type
+ * @param numPartitions user hint on how many RDD partitions to create instead of aligning with Kafka partitions
+ * @param cSharpTransformFunc user function to process the kafka data
+ */
+private[streaming]
+class ParallelKafkaDataFetcherDStream[
+  K: ClassTag,
+  V: ClassTag,
+  U <: Decoder[K]: ClassTag,
+  T <: Decoder[V]: ClassTag,
+  R: ClassTag](
+    @transient _ssc : StreamingContext,
+    val kafkaParams: Map[String, String],
+    topics: Set[String],
+    val fromOffsets: Map[TopicAndPartition, Long],
+    messageHandler: MessageAndMetadata[K, V] => R,
+    numPartitions: Int,
+    cSharpTransformFunc: Array[Byte]
+  ) extends InputDStream[R](_ssc) {
+
+  private val serializationMode = "Pair"
+
+  private val refreshOffsetsInterval = context.sparkContext.getConf.getLong(
```

Comment:
documentation updates needed for new config parameters @ https://github.com/Microsoft/Mobius/blob/master/notes/configuration-mobius.md